### PR TITLE
`kubectl get` and `kubectl describe` extensions

### PR DESCRIPTION
**Pull request by pwittrock**
…ns for non-compiled in types.

@fabianofranz The sig-service-catalog was going to fork kubectl and compile in custom types for get and describe to work.  I really don't like this solution, and think we could support their needs in a simple way.

Would you be open to discussing this on Wednesday?

sig-service-catalog has agreed to donate the development resources to implement.  @simonleung8

### CODE REVIEW

**Comment by pwittrock** [comment]
FYI: @duglin

**Comment by duglin** [comment]
thanks @pwittrock!

**Comment by deads2k** [comment]
get and describe function based on resource builders (as distinct from create).  If you allow people to claim a name under `kubectl get`, you'll effectively "steal" that name from anyone who adds an aggregated API server with that resource.  I think that would be a bad outcome.

**Comment by philips** [comment]
Question I would like to see answered: How does this behavior play with third-party resources and the naming scheme there? Which takes precedence?

**Comment by deads2k** [comment]
> Question I would like to see answered: How does this behavior play with third-party resources and the naming scheme there? Which takes precedence?

Currently, thirdpartyresources are not special.  They are exposed via discovery and the `kubectl` resourcebuilder inspects discovery to find mappable names (with some significant priority and autocompletion rules).  If we were allow extensions under `kubectl get` in the CLI, someone could claim `kubectl get operator` and any TPR named `operator` would be screwed.

**Comment by philips** [comment]
@deads2k Well, that is partially true as you can have two third party resources that are both called "operator". In that case someone can still use `kubectl get operator.third-party.example.com/v1` to find their specific resource.

**Comment by deads2k** [comment]
> @deads2k Well, that is partially true as you can have two third party resources that are both called "operator". In that case someone can still use kubectl get operator.third-party.example.com/v1 to find their specific resource.

True, but in practice claiming a section of the value-space via extension like this feels like cheating.  We wouldn't let people do it in tree because of name stealing, it doesn't seem like we'd want to allow it from out of tree.

**Comment by liggitt** [comment]
Making the generic kubectl operations work consistently via discovery puts control of precedence for a particular server in the cluster-admin's hands. Two users running `kubectl get foo` against the same server and getting different objects seems like a bad idea.

File: contributors/design-proposals/sig-cli/get-describe-apiserver-extensions.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+for resources retrieved through federated apiservers and types not
+compiled into the kubectl binary.  Kubectl should support printing
+columns configured per-type without having the types compiled in.
+
+## Proposal
+
+Allow the apiserver to define the type specific columns that will be
+printed using the open-api swagger.json spec already fetched by kubectl.
+This provides a limited describe to only print out fields on the object
+and related events.
```
Comment (liggitt) [inline contributors/design-proposals/sig-cli/get-describe-apiserver-extensions.md]: how do you envision this interacting with https://github.com/kubernetes/kubernetes/pull/33900 ?

File: contributors/design-proposals/sig-cli/get-describe-apiserver-extensions.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  resources defined in a federated apiserver, I want the command
+  to succeed, and to see events for the resource along with important
+  fields of the resource.
+
+## Implementation
+
+Define the open-api extensions `x-kubernetes-kubectl-get-columns` and
+`x-kubernetes-kubectl-describe-columns`.  These extensions have a
+string value containing the columns to be printed by kubectl.  The
+string format is the same as the `--custom-columns` for `kubectl get`.
```
Comment (liggitt) [inline contributors/design-proposals/sig-cli/get-describe-apiserver-extensions.md]: we don't have openapi schemas for third party resources... do we anticipate adding that in the future?

File: contributors/design-proposals/sig-cli/get-describe-apiserver-extensions.md (hunk 1 above)
Comment (pwittrock) [inline contributors/design-proposals/sig-cli/get-describe-apiserver-extensions.md]: Good question.  I guess I don't fully understand how that proposal is intended to interact with the swagger schema.   I suspect that this proposal is a much lighter version of the sub-pieces of kubernetes/kubernetes#33900 dealing with getting metadata about objects.

File: contributors/design-proposals/sig-cli/get-describe-apiserver-extensions.md (hunk 2 above)
Comment (pwittrock) [inline contributors/design-proposals/sig-cli/get-describe-apiserver-extensions.md]: Thanks for pointing that out.  I just added some comments about this.  The short answer is - Yes I would like to add support for TPR.  My initial thought is that this can be done through an annotation on the TPR.  If we implement this proposal by hooking into the RestMapper, supporting TPR should be relatively clean.  If we hack this proposal in, then it would be less clean.

**Comment by pwittrock** [comment]
@deads2k @philips 

PTAL.  I have added sections to the implementation part that should clarify things.

TL;DR:

@philips TPRs are not being targeted for the alpha version since they will never appear in the swagger.json, but their future integration is being considered as part of the implementation considerations.

@deads2k I wasn't planning on overriding the resource builder, but instead either integrating with it by having it parse the extensions when present.

**Comment by liggitt** [comment]
> integrating with it by having it parse the extensions when present.

would you envision that happening at the two spots we print objects we don't otherwise know about?

https://github.com/kubernetes/kubernetes/blob/e0337b1d0c8877e899888325cc3818cee600e6dd/pkg/kubectl/resource_printer.go#L2431-L2454

and

https://github.com/kubernetes/kubernetes/blob/e0337b1d0c8877e899888325cc3818cee600e6dd/pkg/kubectl/describe.go#L188-L214

**Comment by pwittrock** [comment]
@liggitt Yes that is what I was thinking.  Not certain of the best way to plumb the data there though.

**Comment by fabianofranz** [comment]
This seems to have some overlapping with https://github.com/kubernetes/community/pull/363.

**Comment by smarterclayton** [comment]
The intent would be that the implementation of this be done on the server side, which simplifies clients (they don't have to look at swagger at all).

File: contributors/design-proposals/sig-cli/get-describe-apiserver-extensions.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+columns configured per-type without having the types compiled in.
+
+## Proposal
+
+Allow the apiserver to define the type specific columns that will be
+printed using the open-api swagger.json spec already fetched by kubectl.
+This provides a limited describe to only print out fields on the object
+and related events.
+
+**Note:** This solution will only work for types compiled into the apiserver
```
Comment (AdoHe) [inline contributors/design-proposals/sig-cli/get-describe-apiserver-extensions.md]: include federated apiservers?

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-cli-proposals @kubernetes/sig-api-machinery-proposals @kubernetes/sig-service-catalog-proposals

**Comment by bgrant0607** [comment]
@pwittrock @fabianofranz @smarterclayton Maybe this should be hashed out in a joint SIG CLI / API machinery meeting? Or choose one of the meetings?

**Comment by bgrant0607** [comment]
In general, I'd like OpenAPI to provide rich, complete, accurate info for clients, but for the most commonly needed functionality to be implemented in the apiserver and controllers (e.g., GC). https://github.com/kubernetes/kubernetes/issues/12143

Example:

Schema validation today is done in the client using OpenAPI, but I'd like to provide this optionally in the server: https://github.com/kubernetes/kubernetes/issues/5889. Multi-platform orchestrators like ManageIQ, Ansible, Puppet, Google Deployment Manager, etc. may choose to use OpenAPI, but simpler K8s-specific clients should be able to request schema validation. We don't want to have to write that in 6+ languages (Go, Python, Java, Javascript, Ruby, PHP, ...).

In this case, as I commented https://github.com/kubernetes/community/pull/363/files#r111605582, I envision resource-specific computation and filtering to happen in the server, and rendering to be done in the client. 

For TPR, I imagine accepting the same OpenAPI format that K8s returns for built-in resources, whether from the primary apiserver or aggregated ones. That would support schema validation, declarative value validation and defaulting, get, describe, diff, merge, apply, edit, explain, external orchestrators, client-library generation, documentation generation, etc. We shouldn't invent yet another schema for enabling those things, other than our types.go-based annotations (currently a combination of tags and comments), until/unless that's replaced by another IDL.

cc @lavalamp @cheftako @enisoc @pwittrock @mbohlool

**Comment by pwittrock** [comment]
Exact swagger tag names TBD
